presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentProvider.java

```diff
@@ -45,14 +47,18 @@ protected EnvironmentProvider(List<EnvironmentExtender> bases)
                 .add(environmentConfig)
                 .build();
 
-        String extendersNames = extenders.stream()
-                .map(Object::getClass)
-                .map(Class::getSimpleName)
-                .collect(joining(", "));
-
-        log.info("Building environment %s with extenders: %s", builder.getEnvironmentName(), extendersNames);
-
-        extenders.forEach(extender -> extender.extendEnvironment(builder));
+        Set<EnvironmentExtender> seen = newSetFromMap(new IdentityHashMap<>());
+        extenders.forEach(extender -> extend(extender, builder, seen));
         return builder;
     }
+
+    private void extend(EnvironmentExtender extender, Environment.Builder builder, Set<EnvironmentExtender> seen)
+    {
+        extender.getDependencies()
+                .forEach(dependencyExtender -> extend(dependencyExtender, builder, seen));
+        if (seen.add(extender)) {
```

Comment:
According to 
https://github.com/prestosql/presto/blob/0816357bb0a6fd6681f8942b94fbc79d9f2c21a8/presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentModule.java#L52-L60
`EnvironmentExtenders` are not bound as singletons. So this probably would not work. Can we somehow ensure (via test for example) proper validation for that?

Reply:
+1

Reply:
So no validataion ? :/